### PR TITLE
[SPARK-24538][SQL] ByteArrayDecimalType support push down to the data sources

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

[ByteArrayDecimalType](https://github.com/apache/spark/blob/e28eb431146bcdcaf02a6f6c406ca30920592a6a/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala#L230) support push down to the data sources.

## How was this patch tested?
unit tests and manual tests.

**manual tests**:
```scala
spark.range(10000000).selectExpr("id", "cast(id as decimal(9)) as d1", "cast(id as decimal(9, 2)) as d2", "cast(id as decimal(18)) as d3", "cast(id as decimal(18, 4)) as d4", "cast(id as decimal(38)) as d5", "cast(id as decimal(38, 18)) as d6").coalesce(1).write.option("parquet.block.size", 1048576).parquet("/tmp/spark/parquet/decimal")
val df = spark.read.parquet("/tmp/spark/parquet/decimal/")
// Only read about 1 MB data
df.filter("d6 = 10000").show
spark.sql("set spark.sql.parquet.filterPushdown=false")
// Read 174.3 MB data
df.filter("d6 = 10000").show
```

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 1]
```diff
@@ -359,6 +359,41 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
     }
   }
 
+  test("filter pushdown - decimal(ByteArrayDecimalType)") {
+    val one = new java.math.BigDecimal(1)
+    val two = new java.math.BigDecimal(2)
+    val three = new java.math.BigDecimal(3)
+    val four = new java.math.BigDecimal(4)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Hm, why don't we follow the style in this file with `implicit class`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 2]
```diff
@@ -62,6 +62,11 @@ private[parquet] class ParquetFilters(pushDownDate: Boolean) {
       (n: String, v: Any) => FilterApi.eq(
         intColumn(n),
         Option(v).map(date => dateToDays(date.asInstanceOf[Date]).asInstanceOf[Integer]).orNull)
+    case decimal: DecimalType if DecimalType.isByteArrayDecimalType(decimal) =>
+      (n: String, v: Any) => FilterApi.eq(
+        binaryColumn(n),
+        Option(v).map(d => Binary.fromReusedByteArray(new Array[Byte](8) ++
+            v.asInstanceOf[java.math.BigDecimal].unscaledValue().toByteArray)).orNull)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: v.asInstanceOf -> d.asInstanceOf?

**Comment by wangyum** [comment]
Jenkins, retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 3]
```diff
@@ -37,6 +39,23 @@ private[parquet] class ParquetFilters(pushDownDate: Boolean) {
     DateTimeUtils.fromJavaDate(date)
   }
 
+  private def decimalToBinary(precision: Int, decimal: JBigDecimal): Binary = {
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: REF: https://github.com/apache/spark/blob/21a7bfd5c324e6c82152229f1394f26afeae771c/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala#L247-L266

**Comment by wangyum** [comment]
Jenkins, retest this please.

**Comment by wangyum** [comment]
Jenkins, retest this please.

**Comment by wangyum** [comment]
Close it because I have implement it in [SPARK-24549](https://issues.apache.org/jira/browse/SPARK-24549).

**Pull request closed without merging**